File: packages/runtime/src/protocol.ts
// The Phase 4 split protocol, browser side. Frame = u32 magic "DNM1" (0x444E4D31 little-endian) | u32 header_len
// | JSON header | payload. Payloads are little-endian fp16 [T, d_model] hidden states (N >= 1) or int32 token
// ids (N = 0). Mirrors server/dianome_server/protocol.py.

export const MAGIC = 0x444e4d31;

export type MessageType = "open" | "opened" | "prefill" | "decode" | "token" | "stats" | "close" | "error" | "ping" | "pong";

export interface Message { type: MessageType; header: Record<string, unknown>; payload: Uint8Array }

export function encodeFrame(type: MessageType, header: Record<string, unknown> = {}, payload: ArrayBufferView | null = null): ArrayBuffer {
  const hb = new TextEncoder().encode(JSON.stringify({ ...header, type }));
  const plen = payload ? payload.byteLength : 0;
  const buf = new ArrayBuffer(8 + hb.byteLength + plen);
  const dv = new DataView(buf);
  dv.setUint32(0, MAGIC, true);
  dv.setUint32(4, hb.byteLength, true);
  const u8 = new Uint8Array(buf);
  u8.set(hb, 8);
  if (payload) u8.set(new Uint8Array(payload.buffer, payload.byteOffset, payload.byteLength), 8 + hb.byteLength);
  return buf;
}

export function decodeFrame(buf: ArrayBuffer): Message {
  const dv = new DataView(buf);
  if (buf.byteLength < 8) throw new Error(`frame too short (${buf.byteLength} bytes)`);
  const magic = dv.getUint32(0, true);
  if (magic !== MAGIC) throw new Error(`bad magic 0x${magic.toString(16)}`);
  const hlen = dv.getUint32(4, true);
  if (8 + hlen > buf.byteLength) throw new Error("header_len exceeds frame");
  const header = JSON.parse(new TextDecoder().decode(new Uint8Array(buf, 8, hlen))) as Record<string, unknown>;
  const type = header.type as MessageType;
  if (typeof type !== "string") throw new Error("header has no type");
  return { type, header, payload: new Uint8Array(buf, 8 + hlen) };
}

export interface TokenMessage { id: number; position: number; busy_ms: number; done: boolean }

/** One session over a WebSocket. Browsers cannot set upgrade headers, so the token travels as `?token=`. */
export class SplitClient {
  private ws: WebSocket | null = null;
  private queue: ((m: Message) => void)[] = [];
  private inbox: Message[] = [];
  private closed: Error | null = null;
  opened: Record<string, unknown> | null = null;

  constructor(readonly url: string, readonly token: string) {}

  async connect(): Promise<void> {
    const u = new URL(this.url);
    u.searchParams.set("token", this.token);
    const ws = new WebSocket(u.toString());
    ws.binaryType = "arraybuffer";
    this.ws = ws;
    await new Promise<void>((resolve, reject) => {
      ws.onopen = () => resolve();
      ws.onerror = () => reject(new Error(`WebSocket to ${this.url} failed`));
    });
    ws.onmessage = (ev) => {
      if (!(ev.data instanceof ArrayBuffer)) return;
      const m = decodeFrame(ev.data);
      const w = this.queue.shift();
      if (w) w(m); else this.inbox.push(m);
    };
    ws.onclose = (ev) => { this.closed = new Error(`WebSocket closed (${ev.code} ${ev.reason})`); for (const w of this.queue.splice(0)) w({ type: "error", header: { code: "closed", message: this.closed.message }, payload: new Uint8Array() }); };
    ws.onerror = () => { /* onclose follows */ };
  }

  private next(): Promise<Message> {
    const m = this.inbox.shift();
    if (m) return Promise.resolve(m);
    if (this.closed) return Promise.reject(this.closed);
    return new Promise((resolve) => this.queue.push(resolve));
  }

  async send(type: MessageType, header: Record<string, unknown> = {}, payload: ArrayBufferView | null = null): Promise<Message> {
    if (!this.ws) throw new Error("not connected");
    this.ws.send(encodeFrame(type, header, payload));
    const m = await this.next();
    if (m.type === "error") throw new Error(`server error ${String(m.header.code)}: ${String(m.header.message)}`);
    return m;
  }

  async open(model: string, N: number, maxCtx = 4096, sampling: Record<string, unknown> = { temperature: 0 }): Promise<Record<string, unknown>> {
    const m = await this.send("open", { model, N, max_ctx: maxCtx, sampling });
    if (m.type !== "opened") throw new Error(`expected opened, got ${m.type}`);
    this.opened = m.header;
    return m.header;
  }

  /** fp16 [T, d_model] bits (N >= 1). */
  async prefill(hidden: Uint16Array, T: number, start = 0): Promise<TokenMessage> {
    const m = await this.send("prefill", { T, positions: [start, start + T] }, hidden);
    if (m.type !== "token") throw new Error(`expected token, got ${m.type}`);
    return m.header as unknown as TokenMessage;
  }

  async decode(hidden: Uint16Array, position: number): Promise<TokenMessage> {
    const m = await this.send("decode", { position }, hidden);
    if (m.type !== "token") throw new Error(`expected token, got ${m.type}`);
    return m.header as unknown as TokenMessage;
  }

  /** int32 token ids [T] (N = 0). */
  async prefillIds(ids: Int32Array, start = 0): Promise<TokenMessage> {
    const m = await this.send("prefill", { T: ids.length, positions: [start, start + ids.length] }, ids);
    if (m.type !== "token") throw new Error(`expected token, got ${m.type}`);
    return m.header as unknown as TokenMessage;
  }

  async decodeId(id: number, position: number): Promise<TokenMessage> {
    const m = await this.send("decode", { position }, new Int32Array([id]));
    if (m.type !== "token") throw new Error(`expected token, got ${m.type}`);
    return m.header as unknown as TokenMessage;
  }

  /** Round-trip time of one ping/pong in ms (allowed before open). */
  async ping(): Promise<number> {
    const t0 = performance.now();
    const m = await this.send("ping", { t: t0 });
    if (m.type !== "pong") throw new Error(`expected pong, got ${m.type}`);
    return performance.now() - t0;
  }

  /** Median of `n` pings. */
  async rtt(n = 5): Promise<number> {
    const s: number[] = [];
    for (let i = 0; i < n; i++) s.push(await this.ping());
    s.sort((a, b) => a - b);
    return s[Math.floor(s.length / 2)]!;
  }

  async stats(): Promise<Record<string, unknown>> { return (await this.send("stats", {})).header; }

  close(): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) { this.ws.send(encodeFrame("close", {})); this.ws.close(); }
    this.ws = null;
  }
}

File: server/dianome_server/protocol.py
"""Binary WebSocket frames.

Frame = u32 magic 0x444E4D31 ("DNM1") | u32 header_len | JSON header | payload.
Both u32 are little-endian, so the first four bytes on the wire are 31 4D 4E 44.
The JSON header always carries `"type"`. Payloads are little-endian, row-major:
fp16 `[T, d_model]` hidden states, or int32 `[T]` token ids when N = 0.

| type    | dir | header                                              | payload                          |
| open    | c→s | {model, N, max_ctx, sampling{temperature,top_p,seed}, logits_topk?} | —                  |
| opened  | s→c | {session, L, d_model, boundary: "input_to_block_N"} | —                                |
| prefill | c→s | {T, positions: [start, end)}                        | fp16 [T, d_model] or int32 [T]   |
| decode  | c→s | {position}                                          | fp16 [1, d_model] or int32 [1]   |
| token   | s→c | {id, position, busy_ms, done, topk_ids?}            | optional fp16 [k] top-k logits   |
| stats   | c→s / s→c | {} / {tokens, busy_seconds, gpu_seconds_per_token} | —                          |
| ping    | c→s | {t?}  (RTT probe; allowed before open)              | —                                |
| pong    | s→c | {t?}  (echoes the ping header)                      | —                                |
| close   | any | {}                                                  | —                                |
| error   | s→c | {code, message}                                     | —                                |
"""

from __future__ import annotations

import dataclasses
import struct
from typing import Any, Optional

import numpy as np
import orjson

MAGIC = 0x444E4D31  # "DNM1" read big-endian; sent as little-endian u32
_HEAD = struct.Struct("<II")

TYPES = ("open", "opened", "prefill", "decode", "token", "stats", "close", "error", "ping", "pong")
REQUIRED: dict[str, tuple[str, ...]] = {
    "open": ("model", "N", "max_ctx"),
    "opened": ("session", "L", "d_model", "boundary"),
    "prefill": ("T", "positions"),
    "decode": ("position",),
    "token": ("id", "position", "busy_ms", "done"),
    "stats": (),
    "ping": (),
    "pong": (),
    "close": (),
    "error": ("code", "message"),
}

HIDDEN_DTYPE = np.dtype("<f2")
IDS_DTYPE = np.dtype("<i4")


class ProtocolError(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


@dataclasses.dataclass
class Message:
    type: str
    header: dict[str, Any]
    payload: bytes = b""

    def __getitem__(self, key: str) -> Any:
        return self.header[key]

    def get(self, key: str, default: Any = None) -> Any:
        return self.header.get(key, default)


def encode(type: str, header: Optional[dict[str, Any]] = None, payload: bytes | memoryview = b"") -> bytes:
    if type not in TYPES:
        raise ProtocolError("bad_type", f"unknown message type {type!r}")
    h = dict(header or {})
    h["type"] = type
    hb = orjson.dumps(h)
    return _HEAD.pack(MAGIC, len(hb)) + hb + bytes(payload)


def decode(frame: bytes | bytearray | memoryview) -> Message:
    frame = memoryview(frame)
    if len(frame) < _HEAD.size:
        raise ProtocolError("bad_frame", f"frame too short ({len(frame)} bytes)")
    magic, hlen = _HEAD.unpack_from(frame, 0)
    if magic != MAGIC:
        raise ProtocolError("bad_magic", f"bad magic 0x{magic:08X}")
    if _HEAD.size + hlen > len(frame):
        raise ProtocolError("bad_frame", "header_len exceeds frame")
    try:
        header = orjson.loads(bytes(frame[_HEAD.size : _HEAD.size + hlen]))
    except orjson.JSONDecodeError as e:
        raise ProtocolError("bad_header", f"header is not JSON: {e}") from e
    if not isinstance(header, dict) or "type" not in header:
        raise ProtocolError("bad_header", "header must be a JSON object with a 'type'")
    t = header["type"]
    if t not in TYPES:
        raise ProtocolError("bad_type", f"unknown message type {t!r}")
    missing = [k for k in REQUIRED[t] if k not in header]
    if missing:
        raise ProtocolError("bad_header", f"{t}: missing {', '.join(missing)}")
    return Message(t, header, bytes(frame[_HEAD.size + hlen :]))


# -- payload helpers ----------------------------------------------------------------


def hidden_to_bytes(x) -> bytes:
    """torch or numpy [T, d] -> little-endian fp16 row-major bytes."""
    a = np.ascontiguousarray(_to_numpy(x), dtype=np.float16).astype(HIDDEN_DTYPE, copy=False)
    if a.ndim != 2:
        raise ProtocolError("bad_payload", f"hidden must be 2-D, got shape {a.shape}")
    return a.tobytes()


def bytes_to_hidden(b: bytes, T: int, d_model: int) -> np.ndarray:
    want = T * d_model * 2
    if len(b) != want:
        raise ProtocolError("bad_payload", f"expected {want} bytes of fp16 [{T}, {d_model}], got {len(b)}")
    return np.frombuffer(b, dtype=HIDDEN_DTYPE).reshape(T, d_model).astype(np.float16, copy=False)


def ids_to_bytes(ids) -> bytes:
    a = np.ascontiguousarray(_to_numpy(ids)).astype(IDS_DTYPE)
    if a.ndim != 1:
        raise ProtocolError("bad_payload", f"ids must be 1-D, got shape {a.shape}")
    return a.tobytes()


def bytes_to_ids(b: bytes, T: int) -> np.ndarray:
    want = T * 4
    if len(b) != want:
        raise ProtocolError("bad_payload", f"expected {want} bytes of int32 [{T}], got {len(b)}")
    return np.frombuffer(b, dtype=IDS_DTYPE).astype(np.int32, copy=False)


def _to_numpy(x):
    if hasattr(x, "detach"):  # torch tensor
        return x.detach().cpu().numpy()
    return np.asarray(x)
